### DISCUSSION TITLE
How to import blend or fbx file into {Data/Models/} ?

### DISCUSSION THREAD

**Discussion by crystaldev3d** [rokups/rbfx#457] [General]
Hello developers, I would like to know how the process of importing models into the engine?
I made a blend file with a cube and placed it in the Data/Assets folder, but then I don't know what to do to get the mdl file of the model from blend file.

**Comment by iSLC** [comment]
Checkout the tools `Source/Tools` folder [AssetImporter](https://github.com/rbfx/rbfx/tree/master/Source/Tools/AssetImporter) is probably the one you're looking for unless you're using the editor. It's a command line utility that can be used to convert third-party model formats (supported by [assimp](https://github.com/assimp/assimp)) to a format supported by the engine.

You need to compile the tool tho.

**Reply by crystaldev3d** [reply]
> Checkout the tools `Source/Tools` folder [AssetImporter](https://github.com/rbfx/rbfx/tree/master/Source/Tools/AssetImporter) is probably the one you're looking for

Thank you, yes I saw that toolkit. but I wanted to use the new import tools in the editor, but for some reason they don't work for me, or I just don't know how they work(

**Comment by eugeneko** [comment]
@crystaldev3d check out Editor Settings/ExternalTools. You need to specify path to blend executable, or (alternatively) add it to the system path.

**Reply by crystaldev3d** [reply]
> check out Editor Settings/ExternalTools. You need to specify path to blend executable, or (alternatively) add it to the system path.

Yes I have seen these settings and I think I set them as needed, but I still can't do anything with the blend file

**Reply by crystaldev3d** [reply]
![picture1](https://user-images.githubusercontent.com/95735962/185744104-eba3f01a-1b98-4f31-8d32-e7d7bdac314f.jpg)

**Reply by eugeneko** [reply]
@crystaldev3d That's... weird, because it totally worked for me just now:
![image](https://user-images.githubusercontent.com/8576192/185765116-e6a5c3c8-5d24-4c17-bc59-ccdbee3cec87.png)
![image](https://user-images.githubusercontent.com/8576192/185765125-81d055bc-5536-4951-aeb9-d2c74d604ad5.png)
Anything in log?

**Reply by crystaldev3d** [reply]
![picture2](https://user-images.githubusercontent.com/95735962/185777234-aaee6de7-1aa6-4cd1-89ca-5ec8862cb1b8.jpg)
logs says, what it can't find blender, but setting are green marked)

**Reply by crystaldev3d** [reply]
![picture3](https://user-images.githubusercontent.com/95735962/185777256-2379d6ce-1c22-4380-a8f5-f19d4184aff1.jpg)
but if I remove blend files from Data/Assets and copy them back to Data/Assets it show me the data of blend files.

**Reply by eugeneko** [reply]
@crystaldev3d I think the problem is that Editor cached failed import. So if you set up paths after adding blend file, it just remembered asset state. Does it sound plausible for you?

**Reply by crystaldev3d** [reply]
@eugeneko maybe so, but then why newly added blend files have the same problem?
*.glb file shows the structure, always.
*.blend file shows the structure only if you run the editor and copy the *.blend file to data/assets.  
Next time you run the editor again the *.blend file structure is not shown. 
So you have to constantly delete/copy to see the file structure.

Besides, the main problem is that it is not possible to Save/Extract *.mdl from these files (glb/blend) into Data/Model/

**Reply by eugeneko** [reply]
@crystaldev3d oh, I have an idea
try swapping these two lines in Project.cpp
![image](https://user-images.githubusercontent.com/8576192/185966843-d7db2ec9-ae85-4fbc-8f0b-dbfd9f52d014.png)

**Reply by crystaldev3d** [reply]
@eugeneko Yes it worked.
However, for the data of the first file to be visible, I had to delete it once and copy it again. 
Now all the data are always visible, even after restarting the editor. it seems that it was a bug?

**Reply by eugeneko** [reply]
@crystaldev3d yep, Editor validated cache before loading external tools settings, so your paths were ignored on the first run.
I always used default paths and didn't notice this.

> Besides, the main problem is that it is not possible to Save/Extract *.mdl from these files (glb/blend) into Data/Model/

It was intentional. Cache is read-only by its nature, so you cannot really change files there or move them.
The natural solution would be to implement simple file copying in Resource Browser.
However, it is recommended to use imported files as is because it will save you manual work on asset iterations.
There are other (and better) solutions, but they all require time to code, and my time is in short supply.

**Comment by luciddream-tsin** [comment]
Ubuntu22.04, The Editor can not recognize the blend file :
![截图 2022-08-24 21-36-09](https://user-images.githubusercontent.com/92364028/186432599-8837b11a-de5d-4b85-bf82-c12ca84d8d00.png)

**Reply by eugeneko** [reply]
It is inconvinient behaviour than I didn't have time to fix.
You *do* need to have configured AssetPipeline.json to execute any kind of operation on assets in the folder (or nested folders).

**Reply by luciddream-tsin** [reply]
okay, I will try to fix it.

**Reply by eugeneko** [reply]
Wait, I did fix that! I just forgot about it because there was another bug.
Editor should actually import models in new projects because it creates `AssetPipeline.json` in `Data/`.
I just messed up filtering and this file was ignored.

**Comment by Erupuran** [comment]
How can I import .mdl files in Blender?
I found some add-ons on GitHub, but they only worked for certain games.
Do I need a dedicated add-on for each game?
By the way, the game I want to edit the .mdl file for is Accel World vs Sword Art Online.

**Reply by eugeneko** [reply]
".mdl" is just an abbreviation for "model", it's not a file format per se, just a naming convention.
So, if you need to edit model files for specific game, you should direct this question to people related to this game.
I know nothing of this game, directly or indirectly.
